File: app/src/main/java/me/ykrank/s1next/data/api/model/Pm.java
package me.ykrank.s1next.data.api.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.github.ykrank.androidtools.guava.Objects;
import com.github.ykrank.androidtools.ui.adapter.model.SameItem;

@JsonIgnoreProperties(ignoreUnknown = true)
public class Pm implements Cloneable, SameItem {
    @JsonProperty("plid")
    private String plId;
    @JsonProperty("pmid")
    private String pmId;
    @JsonProperty("pmtype")
    private String pmType;
    @JsonProperty("authorid")
    private String authorId;
    private String author;
    private String subject;
    private String message;
    private long dateline;
    @JsonProperty("msgfromid")
    private String msgFromId;
    @JsonProperty("msgfrom")
    private String msgFrom;
    @JsonProperty("msgtoid")
    private String msgToId;
    //no this data in api, should set it manual
    @JsonIgnore
    private String msgTo;

    public String getPlId() {
        return plId;
    }

    public void setPlId(String plId) {
        this.plId = plId;
    }

    public String getPmId() {
        return pmId;
    }

    public void setPmId(String pmId) {
        this.pmId = pmId;
    }

    public String getPmType() {
        return pmType;
    }

    public void setPmType(String pmType) {
        this.pmType = pmType;
    }

    public String getAuthorId() {
        return authorId;
    }

    public void setAuthorId(String authorId) {
        this.authorId = authorId;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public String getSubject() {
        return subject;
    }

    public void setSubject(String subject) {
        this.subject = subject;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    public long getDateline() {
        return dateline;
    }

    public void setDateline(long dateline) {
        this.dateline = dateline;
    }

    public String getMsgFromId() {
        return msgFromId;
    }

    public void setMsgFromId(String msgFromId) {
        this.msgFromId = msgFromId;
    }

    public String getMsgFrom() {
        return msgFrom;
    }

    public void setMsgFrom(String msgFrom) {
        this.msgFrom = msgFrom;
    }

    public String getMsgToId() {
        return msgToId;
    }

    public void setMsgToId(String msgToId) {
        this.msgToId = msgToId;
    }

    public String getMsgTo() {
        return msgTo;
    }

    public void setMsgTo(String msgTo) {
        this.msgTo = msgTo;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Pm pm = (Pm) o;
        return Objects.equal(plId, pm.plId) &&
                Objects.equal(pmId, pm.pmId) &&
                Objects.equal(pmType, pm.pmType) &&
                Objects.equal(authorId, pm.authorId) &&
                Objects.equal(author, pm.author) &&
                Objects.equal(subject, pm.subject) &&
                Objects.equal(message, pm.message) &&
                Objects.equal(dateline, pm.dateline) &&
                Objects.equal(msgFromId, pm.msgFromId) &&
                Objects.equal(msgFrom, pm.msgFrom) &&
                Objects.equal(msgToId, pm.msgToId) &&
                Objects.equal(msgTo, pm.msgTo);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(plId, pmId, pmType, authorId, author, subject, message, dateline, msgFromId, msgFrom, msgToId, msgTo);
    }

    @Override
    public boolean isSameItem(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Pm pm = (Pm) o;
        return Objects.equal(plId, pm.plId) &&
                Objects.equal(pmId, pm.pmId);
    }
}

File: app/src/main/java/me/ykrank/s1next/data/api/model/Pm.kt
package me.ykrank.s1next.data.api.model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.github.ykrank.androidtools.ui.adapter.StableIdModel
import com.github.ykrank.androidtools.ui.adapter.model.DiffSameItem

@JsonIgnoreProperties(ignoreUnknown = true)
class Pm : Cloneable, DiffSameItem, StableIdModel {
    @JsonProperty("plid")
    var plId: String? = null
    @JsonProperty("pmid")
    var pmId: String? = null
    @JsonProperty("pmtype")
    var pmType: String? = null
    @JsonProperty("authorid")
    var authorId: String? = null
    var author: String? = null
    var subject: String? = null
    var message: String? = null
    var dateline: Long = 0
    @JsonProperty("msgfromid")
    var msgFromId: String? = null
    @JsonProperty("msgfrom")
    var msgFrom: String? = null
    @JsonProperty("msgtoid")
    var msgToId: String? = null
    //no this data in api, should set it manual
    @JsonIgnore
    var msgTo: String? = null

    override val stableId: Long
        get() = pmId?.toLongOrNull() ?: 0


    override fun isSameItem(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Pm

        if (plId != other.plId) return false
        if (pmId != other.pmId) return false

        return true
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Pm

        if (plId != other.plId) return false
        if (pmId != other.pmId) return false
        if (pmType != other.pmType) return false
        if (authorId != other.authorId) return false
        if (author != other.author) return false
        if (subject != other.subject) return false
        if (message != other.message) return false
        if (dateline != other.dateline) return false
        if (msgFromId != other.msgFromId) return false
        if (msgFrom != other.msgFrom) return false
        if (msgToId != other.msgToId) return false
        if (msgTo != other.msgTo) return false

        return true
    }

    override fun hashCode(): Int {
        var result = plId?.hashCode() ?: 0
        result = 31 * result + (pmId?.hashCode() ?: 0)
        result = 31 * result + (pmType?.hashCode() ?: 0)
        result = 31 * result + (authorId?.hashCode() ?: 0)
        result = 31 * result + (author?.hashCode() ?: 0)
        result = 31 * result + (subject?.hashCode() ?: 0)
        result = 31 * result + (message?.hashCode() ?: 0)
        result = 31 * result + dateline.hashCode()
        result = 31 * result + (msgFromId?.hashCode() ?: 0)
        result = 31 * result + (msgFrom?.hashCode() ?: 0)
        result = 31 * result + (msgToId?.hashCode() ?: 0)
        result = 31 * result + (msgTo?.hashCode() ?: 0)
        return result
    }
}
